### Описание данных
**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Содержание


1.   [Подготовка к выгрузке таблиц](#step1)
2.   [Обзор таблиц](#step2)
3.   [Задания](#step3)
  * [Посчитайте, сколько книг вышло после 1 января 2000 года;](#step3.1)
  * [Для каждой книги посчитайте количество обзоров и среднюю оценку;](#step3.2)
  * [Определите издательство, которое выпустило наибольшее число книг толще 50 страниц;](#step3.3)
  * [Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;](#step3.4)
  * [Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.](#step3.5)



## Подготовка к выгрузке таблиц  <a id='step1'></a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': ,
             'pwd': ,
             'host': ,
             'port': , 
             'db': 
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],\
                                                         db_config['pwd'],\
                                                         db_config['host'],\
                                                         db_config['port'],\
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
query_books = ''' SELECT *
            FROM books
        '''
query_authors = ''' SELECT *
            FROM authors
        '''
query_ratings = ''' SELECT *
            FROM ratings
        '''
query_reviews = ''' SELECT *
            FROM reviews
        '''
query_publishers = ''' SELECT *
            FROM books
        '''

## Обзор таблиц  <a id='step2'></a>

In [4]:
books = pd.io.sql.read_sql(query_books, con = engine)
authors = pd.io.sql.read_sql(query_authors, con = engine)
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

In [5]:
tables = [books, authors, ratings, reviews, publishers]
#чтобы было понятно какая таблица вывелась
names_tables = ['books', 'authors', 'ratings', 'reviews', 'publishers']

In [6]:
for i in range(len(tables)):
    print('---------- Table ' + names_tables[i], '-----------------------')
    display(tables[i].head())
    print(tables[i].info())

---------- Table books -----------------------


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
---------- Table authors -----------------------


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
---------- Table ratings -----------------------


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
---------- Table reviews -----------------------


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
---------- Table publishers -----------------------


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


Таблицы связаны друг с другом таблицей **books**, что удобно для работы. Из можно соеднять друг с другом также по полю book_id, кроме таблицы с авторами. Пропусков нет ни в одной таблице.

## Задачи  <a id='step3'></a>

### Посчитайте, сколько книг вышло после 1 января 2000 года  <a id='step3.1'></a>

In [7]:
query1 = '''
            SELECT COUNT(*) AS cnt_books
            FROM books
            WHERE CAST(publication_date AS date) >= '2000-01-01'
'''

In [8]:
cnt = pd.io.sql.read_sql(query1, con = engine)
cnt

cnt_books
0        821

Из 1000 книг, которые у нас есть 821 шт выпустили после 1 января 2000 года.

### Для каждой книги посчитайте количество обзоров и среднюю оценку  <a id='step3.2'></a>

In [9]:
query2 = '''SELECT books.book_id, books.title, COUNT(reviews.review_id) AS amount_review, avg_rat.avg_rating
            FROM books
            LEFT JOIN reviews ON reviews.book_id = books.book_id
            LEFT JOIN (
              SELECT  book_id, ROUND(AVG(rating), 1) AS avg_rating
              FROM ratings
              GROUP BY book_id
            ) AS avg_rat ON avg_rat.book_id = books.book_id
            GROUP BY books.book_id, books.title, avg_rat.avg_rating
            ORDER BY amount_review DESC;
'''

In [10]:
cnt_rew = pd.io.sql.read_sql(query2, con = engine)
cnt_rew.head(10)

book_id                                              title  amount_review  \
0      948                            Twilight (Twilight  #1)              7   
1      627                                      The Alchemist              6   
2      779  The Lightning Thief (Percy Jackson and the Oly...              6   
3      734                                   The Glass Castle              6   
4      207                                    Eat  Pray  Love              6   
5      302  Harry Potter and the Prisoner of Azkaban (Harr...              6   
6      299  Harry Potter and the Chamber of Secrets (Harry...              6   
7      695  The Curious Incident of the Dog in the Night-Time              6   
8      673                             The Catcher in the Rye              6   
9      696             The Da Vinci Code (Robert Langdon  #2)              6   

   avg_rating  
0         3.7  
1         3.8  
2         4.1  
3         4.2  
4         3.4  
5         4.4  
6         4.3  
7         4.1  
8         3.8  
9         3.8

Как видим, большое количество обзоров на книги не обеспечивает высокие оценки. Максимальное количество обзоров = 7 шт. Оценки находятся в пределах от 3.4 до 4.4.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры  <a id='step3.3'></a>

In [11]:
query3 = '''
            SELECT publishers.publisher, COUNT(books.book_id) AS amoun_books
            FROM books
            LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id 
            WHERE books.num_pages > 50     
            GROUP BY publishers.publisher
            ORDER BY amoun_books DESC
            LIMIT 3;
'''

In [12]:
publ = pd.io.sql.read_sql(query3, con = engine)
publ.head()

publisher  amoun_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

**Penguin Books** издательство, которое выпустило больше всего книг(не брошюр), на его счету 42 выпущенные книги за все время. Издательство **Vintage** на 10 шт. меньше выпустило книг 31 шт, а **Grand Central Publishing** — 25 шт.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками  <a id='step3.4'></a>

In [21]:
query4 = '''
            SELECT authors.author, ROUND(AVG(ratings.rating), 2) AS avg_rating
            FROM books 
            LEFT JOIN authors ON authors.author_id = books.author_id
            LEFT JOIN ratings ON ratings.book_id = books.book_id
            WHERE books.book_id IN
                (SELECT book_id
                FROM ratings
                GROUP BY book_id
                HAVING COUNT(rating_id) >= 50)
            GROUP BY authors.author
            ORDER BY avg_rating DESC
            LIMIT 3;
''' 

In [22]:
aut = pd.io.sql.read_sql(query4, con = engine)
aut

author  avg_rating
0         J.K. Rowling/Mary GrandPré        4.29
1  Markus Zusak/Cao Xuân Việt Khương        4.26
2                     J.R.R. Tolkien        4.25

Автором с самой высокой средней оценкой книги **J.K. Rowling/Mary GrandPré**, 4.29 средняя оценка.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок  <a id='step3.5'></a>

In [ ]:
query5 = '''SELECT ROUND(AVG(tab_50.cnt_reviews), 1) AS avg_count_rating
            FROM
                (SELECT COUNT(review_id) AS cnt_reviews
                FROM reviews
                WHERE username IN
                        (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating_id) >= 50) 
                GROUP BY username) AS tab_50;
'''

In [ ]:
users_rews = pd.io.sql.read_sql(query5, con = engine)
users_rews.head()

Пользователи очень активные. В среднем самые активные из читателей пишуть почти 24 отзывов.